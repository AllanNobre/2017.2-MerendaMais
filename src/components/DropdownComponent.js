import React from 'react';
import PropTypes from 'prop-types';
import { View, Picker } from 'react-native';
import styles from '../Styles/GeneralStyles';

const DropdownComponent = props => (
  <View
    style={styles.InputFieldDropdown}
  >
    <Picker
      onValueChange={checkedAttribute => props.callback(checkedAttribute)}
      selectedValue={props.selectedValue}
    >
      {props.pickerTitle}
      {props.pickerBody}

    </Picker>
  </View>
);

DropdownComponent.propTypes = {
  selectedValue: PropTypes.string.isRequired,
  callback: PropTypes.func.isRequired,
  pickerTitle: PropTypes.element.isRequired,
  pickerBody: PropTypes.element.isRequired,
};

export default DropdownComponent;
