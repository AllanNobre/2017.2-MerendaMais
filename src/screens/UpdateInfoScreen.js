import React, { PropTypes } from 'react';
import { MaterialIcons } from '@expo/vector-icons';
import { StyleSheet,
  TouchableOpacity,
  Text,
  TextInput,
  View,
  Alert,
  Picker,
  ScrollView } from 'react-native';
import { TITULAR_COUNSELOR,
  SURROGATE_COUNSELOR } from '../constants';
import Header from '../components/Header';
import { logInfo } from '../../logConfig/loggers';

const FILE_NAME = 'UpdateInfoScreen.js';

const styles = StyleSheet.create({
  principal: {
    flex: 1,
    backgroundColor: 'white',
  },

  content: {
    flex: 6,
    marginTop: 8,
    paddingHorizontal: 15,
  },

  buttonContainer: {
    paddingVertical: 10,
    borderWidth: 1,
    borderRadius: 7,
    marginHorizontal: 15,
    marginTop: 30,
    marginBottom: 20,
    backgroundColor: '#FF9500',
    justifyContent: 'flex-end',
  },

  buttonText: {
    textAlign: 'center',
    color: '#FFF',
  },

  InputFieldStyle: {
    padding: 8,
    marginTop: 1,
    backgroundColor: '#FAFAFA',
    flexDirection: 'row',
    justifyContent: 'flex-start',
    alignItems: 'center',
    borderRadius: 7,
    marginBottom: 10,
    borderWidth: 1,
    borderColor: 'gray',
  },

  InputFieldDropdown: {
    marginTop: 1,
    borderColor: 'gray',
    borderWidth: 1,
    borderRadius: 7,
    marginBottom: 10,
  },

  icon: {
    margin: 5,
  },

  InputStyle: {
    flex: 1,
  },
});

export default class UpdateInfoScreen extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      name: this.props.counselor.name,
      phone: this.props.counselor.profile.phone,
      counselorType: this.props.counselor.profile.counselorType,
    };

    this.validateName = this.validateName.bind(this);
    this.validatePhone = this.validatePhone.bind(this);
  }

  validateName(name) {
    const validName = name.replace(/[^A-Za-záàâãéèêíïóôõöúçñÁÀÂÃÉÈÍÏÓÔÕÖÚÇÑ ]/g, '');
    this.setState({ name: validName });
  }

  validatePhone(phone) {
    const validPhone = phone.replace(/[^0-9]/g, '');
    this.setState({ phone: validPhone });
  }

  register() {
    const nameRegex = /[A-Za-záàâãéèêíïóôõöúçñÁÀÂÃÉÈÍÏÓÔÕÖÚÇÑ ]/g;
    const phoneRegex1 = /[0-9]{11}/g;
    const phoneRegex2 = /[0-9]{10}/g;
    let error = false;
    let errorMessage = '';

    // Validating Counselor Name.
    if (!nameRegex.test(this.state.name) || this.state.name.trim() === '') {
      error = true;
      errorMessage += 'Nome inválido\n';
    }

    // Validating Counselor Phone.
    if (!phoneRegex1.test(this.state.phone) && !phoneRegex2.test(this.state.phone)) {
      error = true;
      errorMessage += 'Telefone inválido\n';
    }

    // Validating Counselor Type.
    if (this.state.counselorType === '') {
      error = true;
      errorMessage += 'Tipo de Conselheiro não selecionado\n';
    }

    // Checking if was found a irregularity in register fields.
    if (error === false) {
      this.props.asyncEditCounselor(this.fetchCounselorData());
    } else {
      Alert.alert('FALHA AO EDITAR DADOS', errorMessage);
    }
  }

  fetchCounselorData() {
    return {
      nuvemCode: this.props.counselor.nuvemCode,
      name: this.state.name,
      token: this.props.counselor.token,
      userName: this.props.counselor.userName,
      profile: {
        cpf: this.props.counselor.profile.cpf,
        phone: this.state.phone,
        isPresident: this.props.counselor.profile.isPresident,
        counselorType: this.state.counselorType,
        segment: this.props.counselor.profile.segment,
        CAE_Type: this.props.counselor.profile.CAE_Type,
        CAE_UF: this.props.counselor.profile.CAE_UF,
        CAE_municipalDistrict: this.props.counselor.profile.CAE_municipalDistrict,
        CAE: this.props.counselor.profile.CAE,
      },
    };
  }

  render() {
    logInfo(FILE_NAME, 'render()',
      `State of update info page: ${JSON.stringify(this.state, null, 2)}`);

    return (

      <View style={styles.principal}>
        <Header
          title={'Editar Informações'}
          backButton
        />
        <ScrollView>
          <View style={styles.content}>

            <Text>Nome</Text>
            <View style={styles.InputFieldStyle}>
              <MaterialIcons name="face" style={styles.icon} size={26} color="black" />
              <TextInput
                style={styles.InputStyle}
                placeholder="Digite o seu nome completo"
                placeholderTextColor="#95a5a6"
                underlineColorAndroid="transparent"
                returnKeyLabel={'next'}
                maxLength={60}
                keyboardType={'default'}
                autoCapitalize={'words'}
                onChangeText={text => this.validateName(text)}
                value={this.state.name}
              />
            </View>

            <Text>Telefone</Text>
            <View style={styles.InputFieldStyle}>
              <MaterialIcons name="phone" style={styles.icon} size={26} color="black" />
              <TextInput
                style={styles.InputStyle}
                placeholder="Digite o seu telefone"
                placeholderTextColor="#95a5a6"
                underlineColorAndroid="transparent"
                returnKeyLabel={'next'}
                maxLength={11}
                keyboardType={'phone-pad'}
                onChangeText={text => this.validatePhone(text)}
                value={this.state.phone}
              />
            </View>

            <Text>Tipo de Conselheiro</Text>
            <View style={styles.InputFieldDropdown}>
              <Picker
                onValueChange={value => this.setState({ counselorType: value })}
                selectedValue={this.state.counselorType}
              >
                <Picker.Item value="" label="Escolha seu cargo" color="#95a5a6" />
                <Picker.Item value={TITULAR_COUNSELOR} label={TITULAR_COUNSELOR} />
                <Picker.Item value={SURROGATE_COUNSELOR} label={SURROGATE_COUNSELOR} />
              </Picker>
            </View>

          </View>
        </ScrollView>
        <TouchableOpacity
          key="infoUpdate"
          style={styles.buttonContainer}
          onPress={() => this.register()}
        >
          <Text style={styles.buttonText}>Concluir</Text>
        </TouchableOpacity>
      </View>
    );
  }
}

const { shape, string, number, bool } = PropTypes;

UpdateInfoScreen.propTypes = {
  asyncEditCounselor: PropTypes.func.isRequired,
  counselor: shape({
    name: string.isRequired,
    nuvemCode: number.isRequired,
    token: string.isRequired,
    userName: string.isRequired,
    profile: shape({
      cpf: string.isRequired,
      phone: string.isRequired,
      isPresident: bool.isRequired,
      counselorType: string.isRequired,
      segment: string.isRequired,
      CAE_Type: string.isRequired,
      CAE_UF: string.isRequired,
      CAE_municipalDistrict: string.isRequired,
      CAE: string.isRequired,
    }).isRequired,
  }).isRequired,
};
