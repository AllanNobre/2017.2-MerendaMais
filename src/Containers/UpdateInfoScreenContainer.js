import { connect } from 'react-redux';
import { asyncEditCounselor } from '../actions/counselorActions';
import UpdateInfoScreen from '../screens/UpdateInfoScreen';

const mapStateToProps = state => (
  {
    counselor: state.counselor,
  }
);

const mapDispatchToProps = dispatch => (
  {
    asyncEditCounselor: async userData => dispatch(await asyncEditCounselor(userData)),
  }
);

export default connect(mapStateToProps, mapDispatchToProps)(UpdateInfoScreen);
