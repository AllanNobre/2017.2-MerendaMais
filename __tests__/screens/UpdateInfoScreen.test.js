import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import configureStore from 'redux-mock-store';
import { TextInput, Picker } from 'react-native';
// imported as a connected component!
import UpdateInfoScreenContainer from '../../src/Containers/UpdateInfoScreenContainer';
import UpdateInfoScreen from '../../src/screens/UpdateInfoScreen';
import NameField from '../../src/components/NameField';
import PhoneField from '../../src/components/PhoneField';
import DropdownComponent from '../../src/components/DropdownComponent';


Enzyme.configure({ adapter: new Adapter() });

const mockStore = configureStore();

const initialState = {
  counselor: {
    nuvemCode: 1,
    email: 'rodolfo@gmail.com',
    name: 'Rodolfo',
    userName: 'rodolfo@gmail.com',
    password: 'senha',
    token: 'tokenGenerico',
    profile: {
      cpf: '12312312312',
      phone: '61991234567',
      isPresident: false,
      counselorType: 'Titular',
      segment: 'Poder executivo',
      CAE_Type: 'Municipal',
      CAE: 'Rio Branco - AC',
      CAE_UF: 'AC - Acre',
      CAE_municipalDistrict: 'Rio Branco',
    },
  },
};

const asyncAction = {
  asyncEditCounselor: () => ({}),
};

const getUpdateInfoWrapper = () => (
  shallow(
    <UpdateInfoScreen
      {...initialState}
      asyncEditCounselor={asyncAction.asyncEditCounselor}
    />,
  )
);

const checkErrorInValidation = (wrapper, spy) => {
  wrapper.instance().updateInformation();
  expect(spy).toHaveBeenCalled();

  expect(wrapper.state('error')).toEqual(true);

  spy.mockClear();
};

const store = mockStore(initialState);

describe('Testing UpdateInfoScreen', () => {
  it('renders as expected', () => {
    const wrapper = shallow(
      <UpdateInfoScreenContainer />,
      { context: { store } },
    ).dive();
    expect(wrapper).toMatchSnapshot();
  });
});

describe('Testing UpdateInfoScreen Input', () => {
  const wrapper = getUpdateInfoWrapper();

  it('should change state when the text of name input component changes', () => {
    const nameInputComponent = wrapper.find(NameField).dive().find(TextInput);
    expect(wrapper.state('name')).toEqual(initialState.counselor.name);
    nameInputComponent.simulate('ChangeText', 'Maria');
    expect(wrapper.state('name')).toEqual('Maria');
  });

  it('should change state when the text of phone input component changes', () => {
    const phoneInputComponent = wrapper.find(PhoneField).dive().find(TextInput);

    expect(wrapper.state('phone')).toEqual(initialState.counselor.profile.phone);
    phoneInputComponent.simulate('ChangeText', '99999999999');
    expect(wrapper.state('phone')).toEqual('99999999999');
  });

  it('should change state when the choice of counselor type input component changes', () => {
    const counselorTypeInputComponent = wrapper.find(DropdownComponent).at(0).dive().find(Picker);
    expect(wrapper.state('counselorType')).toEqual(initialState.counselor.profile.counselorType);
    counselorTypeInputComponent.simulate('valueChange', 'Suplente');
    expect(wrapper.state('counselorType')).toEqual('Suplente');
  });

  it('should change state when the choice of counselor segment input component changes', () => {
    const segmentInputComponent = wrapper.find(DropdownComponent).at(1).dive().find(Picker);
    expect(wrapper.state('segment')).toEqual(initialState.counselor.profile.segment);
    segmentInputComponent.simulate('valueChange', 'Trabalhadores da Educação');
    expect(wrapper.state('segment')).toEqual('Trabalhadores da Educação');
  });
});

describe('Testing UpdateInfoScreen button', () => {
  it('Test if updateInfo Button is rendered', () => {
    const wrapper = getUpdateInfoWrapper();

    const button = wrapper.findWhere(c => c.key() === 'infoUpdate');
    expect(button.length).toEqual(1);
    button.simulate('press');
  });
});

describe('Testing UpdateInfoScreen updateInformation method', () => {
  const wrapper = getUpdateInfoWrapper();
  const spy = jest.spyOn(UpdateInfoScreen.prototype, 'updateInformation');

  it('Test if validation is called', () => {
    wrapper.instance().updateInformation();

    expect(spy).toHaveBeenCalled();
    spy.mockClear();
  });

  it('Test if validation for a empty name input is working', () => {
    wrapper.setState({
      name: '',
      error: false,
    });

    // Here is expected error equal true because a error ocurred on name input.
    checkErrorInValidation(wrapper, spy);
  });

  it('Test if validation for a invalid name input is working', () => {
    wrapper.setState({
      name: '#',
      error: false,
    });

    // Here is expected error equal true because a error ocurred on name input.
    checkErrorInValidation(wrapper, spy);
  });

  it('Test if validation for a invaid phone length input is working', () => {
    wrapper.setState({
      name: 'Rodrigson',
      phone: '123',
      error: false,
    });

    // Here is expected true because a error ocurred on phone input.
    checkErrorInValidation(wrapper, spy);
  });

  it('Test if validation for a invalid phone input is working', () => {
    wrapper.setState({
      name: 'Rodrigson',
      phone: 'Abc#,',
      error: false,
    });

    // Here is expected true because a error ocurred on phone input.
    checkErrorInValidation(wrapper, spy);
  });

  it('Test if validation for a empty counselor type input is working', () => {
    wrapper.setState({
      name: 'Rodrigson',
      phone: '61969696966',
      counselorType: '',
      error: false,
    });

    // Here is expected true because a error ocurred on counselor type input.
    checkErrorInValidation(wrapper, spy);
  });

  it('Test if validation for a empty counselor segment input is working', () => {
    wrapper.setState({
      name: 'Rodrigson',
      phone: '61969696966',
      counselorType: 'Titular',
      segment: '',
      error: false,
    });

    // Here is expected true because a error ocurred on counselor segment input.
    checkErrorInValidation(wrapper, spy);
  });
});
