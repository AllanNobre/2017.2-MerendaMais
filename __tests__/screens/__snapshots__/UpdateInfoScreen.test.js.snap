// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing UpdateInfoScreen renders as expected 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <Header
    backButton={true}
    subTitle=""
    title="Editar Informações"
  />
  <View
    style={
      Object {
        "flex": 6,
        "flexDirection": "column",
        "marginBottom": 9,
        "marginTop": 30,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#FAFAFA",
          "borderColor": "gray",
          "borderRadius": 7,
          "borderWidth": 1,
          "flexDirection": "row",
          "justifyContent": "flex-start",
          "marginBottom": 30,
          "marginHorizontal": 15,
          "paddingLeft": 2,
          "paddingRight": 4,
          "paddingVertical": 10,
        }
      }
    >
      <Icon
        allowFontScaling={false}
        color="black"
        name="face"
        size={32}
        style={
          Object {
            "margin": 5,
          }
        }
      />
      <TextInput
        keyboardType="default"
        maxLength={60}
        onChangeText={[Function]}
        placeholder="Digite seu nome"
        underlineColorAndroid="transparent"
        value="Rodolfo"
        width={280}
      />
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#FAFAFA",
          "borderColor": "gray",
          "borderRadius": 7,
          "borderWidth": 1,
          "flexDirection": "row",
          "justifyContent": "flex-start",
          "marginBottom": 30,
          "marginHorizontal": 15,
          "paddingLeft": 2,
          "paddingRight": 4,
          "paddingVertical": 10,
        }
      }
    >
      <Icon
        allowFontScaling={false}
        color="black"
        name="phone"
        size={32}
        style={
          Object {
            "margin": 5,
          }
        }
      />
      <TextInput
        keyboardType="phone-pad"
        maxLength={11}
        onChangeText={[Function]}
        placeholder="(00)00000-0000"
        underlineColorAndroid="transparent"
        value="6196661234"
        width={280}
      />
    </View>
  </View>
  <TouchableOpacity
    activeOpacity={0.2}
    key="infoUpdate"
    onPress={[Function]}
    style={
      Object {
        "backgroundColor": "#FF9500",
        "borderRadius": 7,
        "borderWidth": 1,
        "justifyContent": "flex-end",
        "marginBottom": 20,
        "marginHorizontal": 15,
        "marginTop": 30,
        "paddingVertical": 10,
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      disabled={false}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#FFF",
          "textAlign": "center",
        }
      }
    >
      Concluir
    </Text>
  </TouchableOpacity>
</View>
`;
